### PR TITLE
Feature/TR-701/Use BrowserList to validate approved browsers

### PR DESCRIPTION
**Pull request by jsconan**
Related to: https://oat-sa.atlassian.net/browse/TR-701

Refactor the way the browser and the operating system are validated by the diagnostic tool.

It now relies on the `browserlist` validation offered by https://github.com/oat-sa/browserslist-app-tao

A list of excluded browsers and operating systems is also offered. This list is managed from the back office, in the settings section:
image

The validation will now work as follow:
- If the browser is present in the default compatibility list, and the version is greater or equal:
  -> the diagnostic says it is compatible, but the support may end soon (status: `success`)
- If the browser is present in the list supplied by the `browserlist` app, and the version is greater or equal:
  -> the diagnostic says it is supported (status: `success`)
- If the browser is present in the exclusion list:
  -> the diagnostic says it is not supported  (status: `warning`)
- If the browser is marked as unsupported in the default list:
  -> the diagnostic says the system needs an update.  (status: `error`).
  _This case is unlikely to happen, unless we modify the default list._
- If the browser is not listed:
  -> the diagnostic says it is not tested  (status: `warning`)

**What to check/test?**
- unit tests must all pass
  - FE
    ```
    cd tao/views/build
    npx grunt connect:dev taoclientdiagnostictest
    ```
  - BE
    ```
    vendor/bin/phpunit taoClientDiagnostic
    ```
- the AC 1 and 2 should be met. AC 3 and 4 were already there, it needs another TAO extension to be  installed: `taoClientRestrict`.

**How to run the application?**
First of all, the extension needs to be installed, obviously:
```
php tao/scripts/installExtension.php taoClientDiagnostic
```

There are actually 2 modes:
- anonymous diagnostic
- logged on diagnostic

To access the diagnostic from the anonymous mode, open this url in your browser:
`http://YOUR_HOST/taoClientDiagnostic/CompatibilityChecker/index`
image

Then click the `Test system compatibility` button.

To access the diagnostic from the logged on mode, first you need to log in, as a user having the role `Readiness checker`.
image

Then open this url in your browser:
`http://YOUR_HOST/taoClientDiagnostic/Diagnostic/index`

A data grid will show first, then you can click the `Launch readiness check` button.
image

The diagnostic part is similar to the anonymous part, with the difference there is a field for the workstation's name.
image

**Note:** By default, not all options nor all testers are enabled. They can be activated from the config file at `config/taoClientDiagnostic/clientDiag.conf.php`.

The testers are all listed in this config and some may be disabled. To enable them all, you need to set the `enabled` property to `true` for all of them.

The following options could also be turned on for a more comprehensive test:
- `requireSchoolName`
- `requireSchoolId`
- `validateSchoolName`

It will add some fields to the diagnostic screen:
image

### CODE REVIEW

File: config/default/CompatibilityChecker.conf.php
Comment (alexeyvy) [inline config/default/CompatibilityChecker.conf.php:19]: just a note, if there's a service without public config options, the config file may be omitted as well as SERVICE_ID and a migration to register it. We could simply extend `ServiceConfigurable` and read the service from ServiceLocator by `::class`:

`$this->getServiceLocator()->get(CompatibilityChecker::class);`

that's just for info, no need to refactor what's done already in this PR

File: controller/ExcludedBrowsers.php
Comment (alexeyvy) [inline controller/ExcludedBrowsers.php:30]: No empty line allowed here as per PSR (same for some other files)

File: model/CompatibilityChecker.php
Comment (alexeyvy) [inline model/CompatibilityChecker.php:55]:
```suggestion
                throw new tao_models_classes_FileNotFoundException('Unable to find the compatibility file');
```

Comment (alexeyvy) [reply model/CompatibilityChecker.php:55]:
```suggestion
                throw new tao_models_classes_FileNotFoundException('Unable to find the compatibility file');
```

File: model/exclusionList/ExclusionListClassService.php
Comment (alexeyvy) [inline model/exclusionList/ExclusionListClassService.php:53]: should we be able to tell empty exclusion lists from uninitialised lists (NULL maybe?)

Same for `$this->excluded`

File: controller/ExcludedBrowsers.php
Comment (alexeyvy) [inline controller/ExcludedBrowsers.php:46]:
```suggestion
```

File: controller/ExcludedOS.php
Comment (alexeyvy) [inline controller/ExcludedOS.php:46]:
```suggestion
```

File: model/exclusionList/ExclusionListClassService.php
Comment (alexeyvy) [inline model/exclusionList/ExclusionListClassService.php:56]: it's redundant as we have proper phpdocs for getInstances signature:

```
     * @return \core_kernel_classes_Resource[]
     */
    public function getInstances($recursive = false, $params = [])
```

Same in the method below

**Review by alexeyvy — COMMENTED** [review]
Great job, @jsconan! I left minor notes, mainly on CS. Still checking this.

- Please place `declare(strict_types = 1);` at the beginning of each new PHP file, some new files are missing that
- I think we will need a throwaway environment for this feature in Playground
- I'm not sure if we have translation in the release pipeline, please be sure po.min files are updated with the UI phrases added in the PR by running `php tao/scripts/taoTranslate.php -v -a=update -e=taoClientDiag` && `php tao/scripts/taoTranslate.php -v -a=compile -e=taoClientDiag` and committing po & po.min files

File: model/CompatibilityChecker.php
Comment (alexeyvy) [inline model/CompatibilityChecker.php:78]: Let's flush a log record on failure with the reason what happened, it is Github down or json_decode error etc

File: model/CompatibilityChecker.php
Comment (alexeyvy) [inline model/CompatibilityChecker.php:100]: `preg_replace('#(\.0+)+($|-)#', '', $ver);` could move to another private method

Comment (jsconan) [reply model/CompatibilityChecker.php]: good catch 👍

File: model/CompatibilityChecker.php
Comment (jsconan) [reply model/CompatibilityChecker.php:89]: Is not the exception caught and logged?
We don't have so many info from the `json_decode` nor the `file_get_contents` unfortunately, to better identify the problem. But I've split the code in order to separate read and decode issues.

**Comment by jsconan** [comment]
Not use I should add the translations in this PR, it will double the number of files for the least. Since this PR will be merged in an integration branch, I suggest doing this before merging the integration branch.

File: model/CompatibilityChecker.php
Comment (alexeyvy) [reply model/CompatibilityChecker.php:89]: in a situation with http error codes in `file_get_contents`, it won't reach our `common_exception_InconsistentData`  having been broken on `file_get_contents()` with a warning, example:

`Warning: file_get_contents(http://gifthub.co): Failed to open stream: HTTP request failed! HTTP/1.1 403 Forbidden
`

`Warning: file_get_contents(http://gifthub.com4): Failed to open stream: php_network_getaddresses: getaddrinfo failed: nodename nor servname provided, or not known in Command line code on line 1`

So the idea is to put file_get_contents in a try-cache and then throw a TAO exception that will get to the log.
It's a bit tricky because those warnings can't be handled unless the function is called in 'suppression mode' (`@` operator).

Let's try this instead of try-cache:

```
$supportedListData = @file_get_contents($supportListUrl);
if ($supportedListData === null) {
    $error = error_get_last();
    throw new tao_models_classes_FileNotFoundException(sprintf(
        'Unable to fetch the list of supported browsers due to error: %s',
        $error['message']
    ));
}

$supportedList = json_decode($supportedListData, true);

if (!$supportedList) {
    throw new common_exception_InconsistentData('Unable to decode list of supported browsers');
}
```

File: controller/CompatibilityChecker.php
Comment (quintanilhar) [inline controller/CompatibilityChecker.php:106]: 👍

File: controller/ExcludedOS.php
Comment (quintanilhar) [inline controller/ExcludedOS.php:45]: I'd recommend combining the 2 conditions into a single if to avoid one more level in the code.
```
if (
    $myForm->isSubmited()
    && $myForm->isValid()
) {}
```

File: controller/ExcludedOS.php
Comment (quintanilhar) [inline controller/ExcludedOS.php:31]: Return type hint is missing:

```suggestion
    public function editInstance(): void
```

File: controller/ExcludedOS.php
Comment (quintanilhar) [inline controller/ExcludedOS.php:65]: You can use return type hint to avoid the comment:

```suggestion
    protected function getClassService(): ExcludedOSClassService
```

File: controller/ExcludedOS.php
Comment (quintanilhar) [inline controller/ExcludedOS.php:35]: Just a small suggestion, I'd avoid using variables to store values that you'll use as arguments for another method only.

```
$myFormContainer = new \tao_actions_form_Instance($this->getCurrentClass(), $this->getCurrentInstance());
```

File: migrations/Version202109201454385081_taoClientDiagnostic.php
Comment (quintanilhar) [inline migrations/Version202109201454385081_taoClientDiagnostic.php:2]: Should we add the license text block here?

File: model/exclusionList/ExcludedBrowserClassService.php
Comment (quintanilhar) [inline model/exclusionList/ExcludedBrowserClassService.php:50]: Can you move this down in the list, please?

The methods are usually grouped based on their visibility:

1- public
2- protected
3- private

File: model/exclusionList/ExclusionListClassService.php
Comment (quintanilhar) [inline model/exclusionList/ExclusionListClassService.php:32]: I think we can remove this package annotation as we have namespaces. ;)

File: model/exclusionList/ExclusionListClassService.php
Comment (quintanilhar) [inline model/exclusionList/ExclusionListClassService.php:67]: For `null` checking is usually better to use strict type comparison.
```suggestion
        if ($this->excluded === null) {
```

File: model/exclusionList/ExclusionListClassService.php
Comment (quintanilhar) [inline model/exclusionList/ExclusionListClassService.php:73]: It seems you got one extra indentation level here.

We use 4 spaces.

**Review by quintanilhar — COMMENTED** [review]
Added some suggestions, feel free to ignore them.

I can approve the PR once you confirm it.

**Review by wazelin — CHANGES_REQUESTED** [review]
Great PR, @jsconan!

Leaving just a few notes on the most noticeable issues.
Will have a closer look later today.

File: controller/ExcludedOS.php
Comment (jsconan) [reply controller/ExcludedOS.php:35]: Good catch, thank you.
However, I think I cannot apply the suggestion fully, since the `$instance` variable is used elsewhere. But I will at least do it for the class.

File: migrations/Version202109201454385081_taoClientDiagnostic.php
Comment (jsconan) [reply migrations/Version202109201454385081_taoClientDiagnostic.php:2]: Good question, usually we don't for the migration scripts, but maybe is it a bad habit. I'll add them.

File: model/exclusionList/ExclusionListClassService.php
Comment (jsconan) [reply model/exclusionList/ExclusionListClassService.php:34]: I agree, actually, this structure is a port of a similar feature supplied by another TAO extension, and I merely reused the same naming conventions... Since this PR is adding these files, better to correct the names now. So I will ;)

File: model/CompatibilityChecker.php
Comment (jsconan) [reply model/CompatibilityChecker.php:111]: I think it was more or less intended, to workaround the weakness of the default comparison made by the native [version_compare](https://www.php.net/manual/en/function.version-compare.php) function. The native function states that `1 < 1.0.0`, instead of the expected `1 == 1.0.0`.
This workaround will states that `1.2-beta == 1.2beta`. While also ensuring than `1-beta == 1.0.0-beta`

File: model/CompatibilityChecker.php
Comment (wazelin) [reply model/CompatibilityChecker.php:101]: The argument unpacking (`...`) has been available since php 5.6 ;)

**Review by quintanilhar — APPROVED** [review]
- [X] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [X] New code is respecting code style rules
- [X] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [X] Acceptance criteria are respected
- [X] Pull request title and description are meaningful
- [X] Pull request's target is not `master`
- [X] Commits are following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [X] Commits messages are meaningful
- [X] Commits are atomic
- [ ] Changelog is updated according to changes (if applicable)
- [ ] Documentation is updated according to changes (if applicable)

**Review by marpesia — APPROVED** [review]
Good job!

**Review Checklist**
- [x] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful
- [ ] Pull request's target is not `master`
- [ ] Commits are following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ]  Commits messages are meaningful
- [ ]  Commits are atomic
- [ ] Changelog is updated according to changes (if applicable)
- [ ] Documentation is updated according to changes (if applicable)

**Pull request merged**
